callbacks/opentelemetry.go

```diff
@@ -0,0 +1,288 @@
+type OpenTelemetryCallbacksHandlerOption func(*OpenTelemetryCallbacksHandlerOptions) *OpenTelemetryCallbacksHandlerOptions
+
+// WithLogPrompts is an option to enable logging of prompts in the OpenTelemetryCallbacksHandler.
+func WithLogPrompts(logPrompts bool) OpenTelemetryCallbacksHandlerOption {
+	return func(opts *OpenTelemetryCallbacksHandlerOptions) *OpenTelemetryCallbacksHandlerOptions {
+		opts.logPrompts = logPrompts
+		return opts
+	}
+}
+
+// WithLogCompletions is an option to enable logging of completions in the OpenTelemetryCallbacksHandler.
+func WithLogCompletions(logCompletions bool) OpenTelemetryCallbacksHandlerOption {
+	return func(opts *OpenTelemetryCallbacksHandlerOptions) *OpenTelemetryCallbacksHandlerOptions {
+		opts.logCompletions = logCompletions
+		return opts
+	}
+}
+
+var _ Handler = &OpenTelemetryCallbacksHandler{}
+
+// HandleAgentAction implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleAgentAction(_ context.Context, _ schema.AgentAction) {
+}
+
+// HandleAgentFinish implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleAgentFinish(_ context.Context, _ schema.AgentFinish) {
+}
+
+// HandleChain implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleChain(ctx context.Context, _ map[string]any, info schema.ChainInfo, next func(ctx context.Context) (map[string]any, error)) (map[string]any, error) {
+	ctx, span := o.tracer.Start(ctx, "langchaingo.chain."+info.Name)
+	defer span.End()
+	return next(ctx)
+}
+
+// HandleChainEnd implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleChainEnd(_ context.Context, _ map[string]any) {
+}
+
+// HandleChainError implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleChainError(_ context.Context, _ error) {
+}
+
+// HandleChainStart implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleChainStart(_ context.Context, _ map[string]any) {
+}
+
+// HandleLLM implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleLLM(ctx context.Context, messages []llms.MessageContent, options llms.CallOptions, next func(ctx context.Context) (*llms.ContentResponse, error)) (*llms.ContentResponse, error) {
+	ctx, span := o.tracer.Start(ctx, "gen_ai.model."+options.Model+".completion")
+	defer span.End()
+	response, err := next(ctx)
+	if err != nil {
+		span.RecordError(err)
+	}
+
+	o.processLLMRequest(span, messages, options)
+	o.processLLMResponse(span, response, options)
+
+	return response, err
+}
+
+// processLLMRequest records information about the request to the span.
+func (o *OpenTelemetryCallbacksHandler) processLLMRequest(span trace.Span, messages []llms.MessageContent, options llms.CallOptions) {
+	o.recordRequestAttributes(options, span)
+
+	if o.opts.logPrompts {
+		o.logPrompts(messages, span)
+	}
+}
+
+// recordRequestAttributes records the request attributes to the span.
+func (*OpenTelemetryCallbacksHandler) recordRequestAttributes(options llms.CallOptions, childSpan trace.Span) {
```

Comment:
Nit, but in the method above, the span comes first, and in this one it comes last. Should the attributes have consistent ordering?

Reply:
- [ ] Thanks, will reorder the arguments for consistency.